samples/net/sockets/http_server/src/server.c

```diff
@@ -0,0 +1,889 @@
+uint64_t get_uptime_ms(void)
+{
+	struct timespec ts;
+
+	if (clock_gettime(CLOCK_MONOTONIC, &ts) == 0) {
+		uint64_t uptime_ms = ts.tv_sec * 1000 + ts.tv_nsec / 1000000;
+		return uptime_ms;
+	} else {
+		return 0;
+	}
+}
+#endif
+
+int http2_server_start(struct http2_server_ctx *ctx, int32_t timeout)
+{
+	printf("\nType 'quit' to quit\n\n");
+	printf("Waiting for incoming connections...\n");
+
+	eventfd_t value;
+
+	int32_t remaining_time = timeout;
+#ifdef __ZEPHYR__
+	int64_t timestamp = k_uptime_get();
+#else
+	int64_t timestamp = get_uptime_ms();
+#endif
+
+	do {
+		if (timeout > 0) {
+#ifdef __ZEPHYR__
+			remaining_time -= (int32_t)k_uptime_delta(&timestamp);
+#else
+			uint64_t current_time = get_uptime_ms();
+
+			remaining_time -= (int32_t)(current_time - timestamp);
+
+#endif
+			if (remaining_time < 0)
+				remaining_time = 0;
+		}
+
+		int ret = poll(ctx->fds, ctx->num_clients + 2, remaining_time);
+
+		if (ret < 0) {
+			LOG_ERR("poll failed");
+			return -errno;
+		}
+
+		for (int i = 0; i < ctx->num_clients + 2; i++) {
+			if (ctx->fds[i].revents & POLLERR) {
+				LOG_ERR("Error on fd %d\n", ctx->fds[i].fd);
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			if (ctx->fds[i].revents & POLLHUP) {
+				LOG_INF("Client on fd %d has disconnected\n",
+					ctx->fds[i].fd);
+				close_client_connection(ctx, i);
+				continue;
+			}
+
+			if (!(ctx->fds[i].revents & POLLIN))
+				continue;
+
+			if (i == 0) {
+				int new_socket =
+					accept_new_client(ctx->server_fd);
+
+				for (int j = 2; j < MAX_CLIENTS + 2; j++) {
+					if (ctx->fds[j].fd != 0)
+						continue;
+
+					ctx->fds[j].fd = new_socket;
+					ctx->fds[j].events = POLLIN;
+					ctx->clients[j - 2].client_fd =
+						new_socket;
+					ctx->clients[j - 2].offset = 0;
+					ctx->clients[j - 2].is_http2 = 0;
+					ctx->clients[j - 2]
+						.streams[0]
+						.stream_state = -1;
+
+					if (j > ctx->num_clients + 1)
+						ctx->num_clients++;
+
+					break;
+				}
+				continue;
+			}
+
+			if (i == 1) {
+#ifdef __ZEPHYR__
+				eventfd_read(ctx->event_fd, &value);
+#else
+				read(ctx->event_fd, &value, sizeof(value));
+#endif
+				printf("Received stop event. exiting ..\n");
+
+				return 0;
+			}
+
+			int valread = recv(ctx->clients[i - 2].client_fd,
+				ctx->clients[i - 2].buffer +
+					ctx->clients[i - 2].offset,
+				sizeof(ctx->clients[i - 2].buffer) -
+					ctx->clients[i - 2].offset,
+				0);
+
+			if (valread < 0) {
+				LOG_ERR("ERROR reading from socket");
+				close_client_connection(ctx, i - 2);
+				continue;
+			}
+
+			if (valread == 0) {
+				LOG_INF("Connection closed by peer.\n");
+				close_client_connection(ctx, i - 2);
+				continue;
+			}
+
+			ctx->clients[i - 2].offset += valread;
+			ctx->clients[i - 2].server_state = INITIAL_STATE;
+			handle_http_request(ctx, &ctx->clients[i - 2], i - 2);
+		}
+	} while (ctx->infinite == 1);
+	return 0;
+}
+
+void close_client_connection(
+	struct http2_server_ctx *ctx_server, int client_index)
```

Comment:
I think it'd be better to accept `struct http2_client_ctx *client` as the second parameter. Instead of doing some not-always-obvious math on the intext, the `http2_client_ctx` could simply remember which FDS slot it uses (i. e. index in the FDS table).